RecoTracker/LSTCore/src/alpaka/Triplet.h

```diff
@@ -206,63 +206,307 @@ namespace lst {
-    } else if (layer1 == 2 and layer2 == 3 and layer3 == 7) {
-      return alpaka::math::abs(acc, residual) < 1.f;
-    } else if (layer1 == 2 and layer2 == 7 and layer3 == 8) {
-      return alpaka::math::abs(acc, residual) < 1.f;
-    } else if (layer1 == 3 and layer2 == 4 and layer3 == 5) {
-      return alpaka::math::abs(acc, residual) < 2.7f;
-    } else if (layer1 == 4 and layer2 == 5 and layer3 == 6) {
-      return alpaka::math::abs(acc, residual) < 3.06f;
-    } else if (layer1 == 7 and layer2 == 8 and layer3 == 9) {
-      return alpaka::math::abs(acc, residual) < 1;
-    } else if (layer1 == 8 and layer2 == 9 and layer3 == 10) {
-      return alpaka::math::abs(acc, residual) < 1;
-    } else if (layer1 == 9 and layer2 == 10 and layer3 == 11) {
-      return alpaka::math::abs(acc, residual) < 1;
-    } else {
-      return alpaka::math::abs(acc, residual) < 5;
+    //get the type of module: 0 is ps, 1 is 2s
+    const int moduleType3 = modulesInGPU.moduleType[outerOuterLowerModuleIndex];
+
+    //get the x,y position of each MD
+    const float x1 = mdsInGPU.anchorX[firstMDIndex] / 100;
+    const float x2 = mdsInGPU.anchorX[secondMDIndex] / 100;
+    const float x3 = mdsInGPU.anchorX[thirdMDIndex] / 100;
+
+    const float y1 = mdsInGPU.anchorY[firstMDIndex] / 100;
+    const float y2 = mdsInGPU.anchorY[secondMDIndex] / 100;
+    const float y3 = mdsInGPU.anchorY[thirdMDIndex] / 100;
+
+    //set initial and target points
+    float x_init = x2;
+    float y_init = y2;
+    float z_init = z2;
+    float r_init = r2;
+
+    float z_target = z3;
+    float r_target = r3;
+
+    float x_other = x1;
+    float y_other = y1;
+    float z_other = z1;
+    float r_other = r1;
+
+    float dz = z2 - z1;
+
+    //use MD2 for regions 5 and 19 because it works better (see https://github.com/SegmentLinking/cmssw/pull/92)
+    if ((layer1 == 8 && layer2 == 14 && layer3 == 15) || (layer1 == 3 && layer2 == 12 && layer3 == 13)){
+      x_init = x1;
+      y_init = y1;
+      z_init = z1;
+      r_init = r1;
+
+      z_target = z2;
+      r_target = r2;
+
+      x_other = x3;
+      y_other = y3;
+      z_other = z3;
+      r_other = r3;
+
+      dz = z3 - z1;
     }
-  };
+
+    //use the 3 MDs to fit a circle. This is the circle parameters, for circle centers and circle radius
+    float x_center = circleCenterX / 100;
+    float y_center = circleCenterY / 100;
+    float pt = 2 * k2Rinv1GeVf * circleRadius; //k2Rinv1GeVf is already in cm^(-1)
+
+    //determine the charge
+    int charge = 0;
+    if ((x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1) > 0)
+      charge = -1;
+    else
+      charge = 1;
+
+    //get the absolute value of px and py at the initial point
+    float cot_phi_abs = alpaka::math::abs(acc, (y_init - y_center) / (x_init - x_center)); //absolute value of cot(phi)
+    float px = pt * cot_phi_abs / alpaka::math::sqrt(acc, 1 + cot_phi_abs * cot_phi_abs);  //px = pt * cos(phi)
```

Comment:
can this and the sign flips below be replaced by using `2 * k2Rinv1GeVf * (y_init - y_center)` ? It seems like a lot of round-trip transformations are used

Reply:
I've simplified the calculation of px and py, but the sign correction is still needed